core/z3solver_reachability.cpp

```diff
@@ -13,86 +13,99 @@ Z3ReachabilityExpression::Z3ReachabilityExpression(ReachabilityExpression reacha
+                ::warning("Solver returned unknown result for %1%.", node);
+                hasChanged = reachabilityAssignment.has_value();
+                continue;
+            }
+            /// There is no way to falsify the condition. It is always true.
+            if (!solverResult.value()) {
+                it.setReachability(true);
+                hasChanged = !reachabilityAssignment.has_value() || !reachabilityAssignment.value();
+                continue;
+            }
 
-        if (solverResult.value() && solverResult.value()) {
-            it->second.setReachability(std::nullopt);
-            hasChanged = reachabilityAssignment.has_value();
+            if (solverResult.value() && solverResult.value()) {
+                it.setReachability(std::nullopt);
+                hasChanged = reachabilityAssignment.has_value();
+            }
         }
-
-        return hasChanged;
     }
+    return hasChanged;
 }
 
 Z3SolverReachabilityMap::Z3SolverReachabilityMap(const ReachabilityMap &map)
     : symbolMap(map.getSymbolMap()) {
     Util::ScopedTimer timer("Precomputing Z3 Reachability");
     Z3Translator z3Translator(solver);
-    for (const auto &[node, reachabilityExpression] : map) {
-        reachabilityExpression.getCondition()->apply(z3Translator);
-        this->insert(
-            {node, Z3ReachabilityExpression(reachabilityExpression, z3Translator.getResult())});
+    for (const auto &[node, reachabilityExpressionVector] : map) {
+        std::vector<Z3ReachabilityExpression> result;
+        for (auto &reachabilityExpression : reachabilityExpressionVector) {
+            reachabilityExpression.getCondition()->apply(z3Translator);
+            result.push_back(
+                Z3ReachabilityExpression(reachabilityExpression, z3Translator.getResult()));
+        }
+        this->insert({node, result});
     }
 }
 
-std::optional<const ReachabilityExpression *> Z3SolverReachabilityMap::getReachabilityExpression(
-    const IR::Node *node) const {
-    auto it = find(node);
-    if (it != end()) {
-        return &it->second;
+std::optional<std::vector<const ReachabilityExpression *>>
+Z3SolverReachabilityMap::getReachabilityExpressions(const IR::Node *node) const {
+    auto vectorIt = find(node);
+    if (vectorIt != end()) {
+        BUG_CHECK(!vectorIt->second.empty(), "Reachability vector for node %1% is empty.", node);
+        std::vector<const ReachabilityExpression *> result;
+        for (auto &it : vectorIt->second) {
+            result.push_back(&it);
+        }
+        return result;
     }
-
     return std::nullopt;
 }
 
 bool Z3SolverReachabilityMap::updateReachabilityAssignment(const IR::Node *node,
                                                            std::optional<bool> reachability) {
-    auto it = find(node);
-    if (it != end()) {
-        it->second.setReachability(reachability);
-        return true;
-    }
+    /// FIXME: Do we want to keep this? I cannot find any usage.
```

Comment:
This should be used by the control plane code, any time you get an update you may have to change that. Maybe I use a different technique now?

Reply:
I think now we just call `recomputeReachability` in [flay_service.cpp](https://github.com/fruffy/flay/blob/master/core/flay_service.cpp#L75).

Reply:
Okay